### PR TITLE
fix 4463 switch canvg to html2canvas

### PR DESCRIPTION
**Pull request by garsuga**
This includes some changes to the vtk screenshot code introduced in a recent PR. html2canvas is considerably slower than canvg when dealing with only an svg so additional optimization was needed to make the vtk screenshot code smooth. I spoke to rob briefly about this but the vtk screenshot code will likely need a better solution in the future to change how it is called

### CODE REVIEW

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 1]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
...
+                    canvasWorking = new Object();
+                    const tempCanvasWorking = canvasWorking;
+                    setTimeout(() => {
+                        if(tempCanvasWorking === canvasWorking) {
+                            processCanvas();
+                        } else {
+                            waitForUserInteractionPause();
+                        }
+                    }, 250);
+                }
```
Comment (mkeilman) [inline sirepo/package_data/static/js/sirepo-plotting-vtk.js:2822]: We have a [debounce](https://github.com/radiasoft/sirepo/blob/fbc1e6473dde18ffbfb2c884ea8b6154c75ca794/sirepo/package_data/static/js/sirepo-components.js#L4773) function that may serve here. Worth a look if it simplifies this.

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 2]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
...
+                        }
+                        context.drawImage(canvas3d, 0, 0, canvas.width, canvas.height);
+                        if (axesCanvas) {
+                            html2canvas(axesCanvas.parentElement,
+                            {
+                                backgroundColor: null
+                            }).then(c => {
+                                context.drawImage(c, 0, 0);
+                                canvasWorking = null;
+                                if(doNextCanvasProcess) {
```
Comment (mkeilman) [inline sirepo/package_data/static/js/sirepo-plotting-vtk.js:2803]: Space after ```if```

**Review by mkeilman — CHANGES_REQUESTED** [review]
Mostly style comments from me

File: sirepo/package_data/static/js/irad.js [hunk 3]
```diff
@@ -3,10 +3,10 @@
...
-    '<div data-ng-switch-when="dicomPlot" data-dicom-plot="" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
-    '<div data-ng-switch-when="dicomComparePlot" data-dicom-plot="" data-is-compare-plot="1" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
-    '<div data-ng-switch-when="doseDifferencePlot" data-dicom-plot="" data-is-difference-plot="1" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
-    '<div data-ng-switch-when="dicom3d" data-dicom-3d="" class="sr-plot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
+    '<div data-ng-switch-when="dicomPlot" data-dicom-plot="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>',
+    '<div data-ng-switch-when="dicomComparePlot" data-dicom-plot="" data-is-compare-plot="1" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>',
+    '<div data-ng-switch-when="doseDifferencePlot" data-dicom-plot="" data-is-difference-plot="1" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>',
+    '<div data-ng-switch-when="dicom3d" data-dicom-3d="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/irad.js]: [0217a9f](https://github.com/radiasoft/sirepo/pull/4498/commits/0217a9f54162c5f58fb206a13763fa4f8338aa66)

File: sirepo/package_data/static/js/jspec.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ SIREPO.app.config(function() {
         'ring-lattice': '.tfs,.txt',
     };
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="rateCalculation" data-rate-calculation-panel="" class="sr-plot"></div>',
+        '<div data-ng-switch-when="rateCalculation" data-rate-calculation-panel="" class="sr-plot sr-screenshot"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/jspec.js]: [0217a9f](https://github.com/radiasoft/sirepo/pull/4498/commits/0217a9f54162c5f58fb206a13763fa4f8338aa66)

File: sirepo/package_data/static/js/irad.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/irad.js]: [0217a9f](https://github.com/radiasoft/sirepo/pull/4498/commits/0217a9f54162c5f58fb206a13763fa4f8338aa66)

File: sirepo/package_data/static/js/jspec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/jspec.js]: [0217a9f](https://github.com/radiasoft/sirepo/pull/4498/commits/0217a9f54162c5f58fb206a13763fa4f8338aa66)

File: sirepo/package_data/static/js/madx.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ SIREPO.app.config(function() {
     SIREPO.PLOTTING_COLOR_MAP = 'afmhot';
     SIREPO.SINGLE_FRAME_ANIMATION = ['twissAnimation', 'twissFromParticlesAnimation'];
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="matchSummaryAnimation" data-match-summary-panel="" class="sr-plot"></div>',
+        '<div data-ng-switch-when="matchSummaryAnimation" data-match-summary-panel="" class="sr-plot sr-screenshot"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/madx.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/madx.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/ml.js [hunk 6]
```diff
@@ -35,8 +35,8 @@ SIREPO.app.config(function() {
         '</div>',
     ].join('');
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="classificationMetrics" data-table-panel="" data-model-name="{{ modelKey }}" class="sr-plot"></div>',
-        '<div data-ng-switch-when="confusionMatrix" data-table-panel="" data-model-name="{{ modelKey }}" class="sr-plot"></div>',
+        '<div data-ng-switch-when="classificationMetrics" data-table-panel="" data-model-name="{{ modelKey }}" class="sr-plot sr-screenshot"></div>',
+        '<div data-ng-switch-when="confusionMatrix" data-table-panel="" data-model-name="{{ modelKey }}" class="sr-plot sr-screenshot"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/ml.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/ml.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/opal.js [hunk 7]
```diff
@@ -27,7 +27,7 @@ SIREPO.app.config(function() {
         '</div>',
     ].join('');
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="beamline3d" data-beamline-3d="" class="sr-plot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
+        '<div data-ng-switch-when="beamline3d" data-beamline-3d="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/opal.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/opal.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/raydata.js [hunk 8]
```diff
@@ -13,7 +13,7 @@ SIREPO.app.config(() => {
         </div>
     `;
     SIREPO.appReportTypes  = `
-        <div data-ng-switch-when="pngImage" data-png-image="" class="sr-plot" data-model-name="{{ modelKey }}"></div>
+        <div data-ng-switch-when="pngImage" data-png-image="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/raydata.js:16]: Already template string

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/raydata.js:16]: Already template string

File: sirepo/package_data/static/js/rs4pi.js [hunk 9]
```diff
@@ -5,7 +5,7 @@ var srdbg = SIREPO.srdbg;
 
 SIREPO.app.config(function() {
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="dicom" data-dicom-plot="" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
+        '<div data-ng-switch-when="dicom" data-dicom-plot="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/rs4pi.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/rs4pi.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/silas.js [hunk 10]
```diff
@@ -23,7 +23,7 @@ SIREPO.app.config(function() {
         '<li data-export-python-link="" data-report-title="{{ reportTitle() }}"></li>',
     ].join('');
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="crystal3d" data-crystal-3d="" class="sr-plot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
+        '<div data-ng-switch-when="crystal3d" data-crystal-3d="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/silas.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/silas.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/srw.js [hunk 11]
```diff
@@ -11,7 +11,7 @@ SIREPO.app.config(function() {
     SIREPO.PLOTTING_COLOR_MAP = 'grayscale';
     SIREPO.PLOTTING_SHOW_FWHM = true;
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="beamline3d" data-beamline-3d="" class="sr-plot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
+        '<div data-ng-switch-when="beamline3d" data-beamline-3d="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/srw.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/srw.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/warpvnd.js [hunk 12]
```diff
@@ -10,8 +10,8 @@ SIREPO.app.config(function() {
     SIREPO.SINGLE_FRAME_ANIMATION = ['optimizerAnimation', 'fieldCalcAnimation', 'fieldComparisonAnimation'];
     SIREPO.appReportTypes = [
         '<div data-ng-switch-when="conductorGrid" data-conductor-grid="" class="sr-plot" data-model-name="{{ modelKey }}" data-report-id="reportId"></div>',
-        '<div data-ng-switch-when="impactDensity" data-impact-density-plot="" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
-        '<div data-ng-switch-when="optimizerPath" data-optimizer-path-plot="" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
+        '<div data-ng-switch-when="impactDensity" data-impact-density-plot="" class="sr-screenshot sr-plot" data-model-name="{{ modelKey }}"></div>',
+        '<div data-ng-switch-when="optimizerPath" data-optimizer-path-plot="" class="sr-screenshot sr-plot" data-model-name="{{ modelKey }}"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/warpvnd.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/warpvnd.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/webcon.js [hunk 13]
```diff
@@ -35,7 +35,7 @@ SIREPO.app.config(function() {
         '</div>',
     ].join('');
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="bpmMonitor" data-bpm-monitor-plot="" class="sr-plot" data-model-name="{{ modelKey }}"></div>',
+        '<div data-ng-switch-when="bpmMonitor" data-bpm-monitor-plot="" class="sr-plot sr-screenshot" data-model-name="{{ modelKey }}"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/webcon.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/webcon.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/zgoubi.js [hunk 14]
```diff
@@ -14,7 +14,7 @@ SIREPO.app.config(function() {
         '</div>',
     ].join('');
     SIREPO.appReportTypes = [
-        '<div data-ng-switch-when="twissSummary" data-twiss-summary-panel="" class="sr-plot"></div>',
+        '<div data-ng-switch-when="twissSummary" data-twiss-summary-panel="" class="sr-plot sr-screenshot"></div>',
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/zgoubi.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/zgoubi.js]: [e8a2c25](https://github.com/radiasoft/sirepo/pull/4498/commits/e8a2c25472ee420d92a2fd9fe8a427494efc62d2)

File: sirepo/package_data/static/js/sirepo-components.js [hunk 15]
```diff
@@ -4245,34 +4222,7 @@ SIREPO.app.service('plotToPNG', function($http) {
...
-                css.type = 'text/css';
-                // work-around bug fix #857, canvg.js doesn't handle non-standard css
-                cssText = cssText.replace('input::-ms-clear', 'ms-clear');
-                css.appendChild(document.createTextNode(cssText));
-                svg.insertBefore(css, svg.firstChild);
-            }
-            downloadPlot(svg, height, plot3dCanvas, fileName);
-        }
-        promises[0].then(cssResponse);
+        downloadPlot(svg, height, plot3dCanvas, fileName);
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-components.js]: [e2b1adf](https://github.com/radiasoft/sirepo/pull/4498/commits/e2b1adf0b49d63cc0e23990d0a0759690ca52837)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-components.js]: [e2b1adf](https://github.com/radiasoft/sirepo/pull/4498/commits/e2b1adf0b49d63cc0e23990d0a0759690ca52837)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 16]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
     }
 
     this.pngCanvas = function(reportId, vtkRenderer, panel) {
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js:2747]: [0842ade](https://github.com/radiasoft/sirepo/pull/4498/commits/0842ade87684c096cf299d06a45478da68416104)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js:2747]: [0842ade](https://github.com/radiasoft/sirepo/pull/4498/commits/0842ade87684c096cf299d06a45478da68416104)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 17]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
...
-                        canvas.getContext('2d').drawSvg(axesCanvas.outerHTML, 0, 0, canvas.width, canvas.height);
-                    }
-                });
+            copyCanvas: (event, doTraverse) => {
+                if(canvasWorking) {
+                    canvasWorking = new Object();
+                    doNextCanvasProcess = () => res.copyCanvas(event, doTraverse);
+                    return;
+                }
+                const processCanvas = () => {
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js (hunk 2 above)
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 18]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
...
+                }
+                const waitForUserInteractionPause = () => {
+                    canvasWorking = new Object();
+                    const tempCanvasWorking = canvasWorking;
+                    setTimeout(() => {
+                        if(tempCanvasWorking === canvasWorking) {
+                            processCanvas();
+                        } else {
+                            waitForUserInteractionPause();
+                        }
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 19]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
...
-                    else {
-                        vtkRenderer.getRenderWindow().render();
-                    }
-                    canvas.getContext('2d').drawImage(canvas3d, 0, 0, canvas.width, canvas.height);
-                    if (axesCanvas) {
-                        canvas.getContext('2d').drawSvg(axesCanvas.outerHTML, 0, 0, canvas.width, canvas.height);
-                    }
-                });
+            copyCanvas: (event, doTraverse) => {
+                if(canvasWorking) {
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [c98b257](https://github.com/radiasoft/sirepo/pull/4498/commits/c98b257906547c80bec2bfd8d696a386ff8a0660)

File: sirepo/package_data/static/js/sirepo-plotting-vtk.js [hunk 20]
```diff
@@ -2767,30 +2767,60 @@ SIREPO.app.service('vtkToPNG', function(panelState, plotToPNG, utilities) {
     }
 
     this.pngCanvas = function(reportId, vtkRenderer, panel) {
+        var canvasWorking = null;
+        var doNextCanvasProcess = null;
+        
         const canvas = document.createElement('canvas');
+        const context = canvas.getContext("2d");
```
[the remark this answers is not in the document]
Comment (garsuga) [reply sirepo/package_data/static/js/sirepo-plotting-vtk.js]: [b79385e](https://github.com/radiasoft/sirepo/pull/4498/commits/b79385e94f5e2a48693a6e8e5fb93f65bf0c4e8a)

**Comment by mkeilman** [comment]
I'll review this later today

**Review by mkeilman — APPROVED** [review]
All looking good!

**Pull request merged**